src/project/select-modules-in-sources.ts

```diff
@@ -359,3 +372,135 @@ function changeModuleDirectiveIfNecessary(
+            return [{ location: indicesArg.loc }];
+        }
+    }
+
+    // does not delete potential commas between the indices. However, indices are usually so long
+    // that there is one per line, and you probably don't use commas then (prettier does not)
+    // it does leave a lot of whitespace, but that's a common problem (and can be fixed by running
+    // prettier after withModuleSelection())
+    return changes;
+}
+
+interface IndexChange {
+    readonly remove?: boolean;
+    readonly changeFields?: ReadonlyArray<string>;
+}
+
+function changeIndexIfNecessary(
+    indexDef: ObjectValueNode,
+    type: RootEntityType,
+    selectedModules: ReadonlySet<string>,
+): IndexChange {
+    const fieldsArg = indexDef.fields.find((f) => f.name.value === 'fields');
+    if (!fieldsArg) {
+        // invalid index - do not change anything
+        return {};
+    }
+    let fields: ReadonlyArray<string>;
+    if (fieldsArg.value.kind === Kind.STRING) {
+        fields = [fieldsArg.value.value];
+    } else if (fieldsArg.value.kind === Kind.LIST) {
+        fields = fieldsArg.value.values.filter((v) => v.kind === Kind.STRING).map((v) => v.value);
+    } else {
+        // invalid index - do not change anything
+        return {};
+    }
+
+    const fieldsSoFar: string[] = [];
+    for (const field of fields) {
+        // it's easier to reconstruct the IndexField instances than using type.indices because
+        // we would need to find the matching IndexField instances
+        const indexField = new IndexField(field, type, undefined);
+        const fieldsInPath = indexField.fieldsInPath;
+        if (!fieldsInPath) {
+            // if there is an error resolving the field paths, we can't really do anything
+            // behave like in other situations here and in doubt do not touch the definition
+            return {};
+        }
+
+        if (fieldsInPath.some((f) => !f.effectiveModuleSpecification.includedIn(selectedModules))) {
+            // The index is not fully covered by the selected modules
+            const uniqueArg = indexDef.fields.find((f) => f.name.value === 'unique');
+            if (uniqueArg && uniqueArg.value.kind === Kind.BOOLEAN && uniqueArg.value.value) {
+                // A unique index that is not fully covered needs to be removed
+                // (if we were to keep a prefix of the index, the unique behavior would change)
+                return { remove: true };
+            }
+
+            if (fieldsSoFar.length) {
+                // for non-unique indices, include a prefix index
+                // TODO only include this prefix index if there isn't already an identical one
+                // (but first discuss whether the prefix-trimming behavior is actually what we want
+                // - it can be confusing because it only applies to non-unique indices)
+                return { changeFields: fieldsSoFar };
+            } else {
+                return { remove: true };
+            }
```

Comment:
Not sure if I completely understand this one.
You are iterating over the fields of one index, but if some segments in the path of one field are not covered you return with all fields encountered in the index so far.
What is with the other fields which follow afterward? Couldn't they be valid fields again that need to be included?

Reply:
it implements this part of the specification

> If a the common model defines an index where only some fields are part of the selected modules, a
> prefix index needs to be present in the solution schema.
> 
> Consider this common model
> 
> ```
> type Delivery @rootEntity(indices: [
>   { fields: ["deliveryNumber"], unique: true }
>   { fields: ["deliveryNumber", "deliveryLineItems.lineItemNumber"] }
>   { fields: ["isDraft", "exportIdenitifer"] }
> ]) {
>   deliveryNumber: String @key @modules(all: true)
>   isDraft: Boolean @modules(all: true)
>   exportIdentifier: String @modules(in: "customs")
>   deliveryLineItems: DeliveryLineItems @modules(in: "shipping")
> }
> ```
> 
> A solution with only the `shipping` module enabled would need to specify the following indices
> 
> -   `["deliveryNumber"]`
> -   `["deliveryNumber", "deliveryLineItems.lineItemNumber"]`
> -   `["isDraft"]` (a prefix of the common model index)

The idea behind this was that the "isDraft, exportIdentifier" index might have a dual use: It can be used in customs-related apps to look up a non-draft delivery by exportIdentifier, and it can be used (non-customs) shipping apps to show drafts in general.

The same dual use would not apply for an "exportIdentifier, isDraft" index - it could only be used* by customs-related apps because the index is only usable if you filter by "exportIdentifier". Therefore, a solution that does not enable customs does not need this index at all.

The whole logic is a bit weird though. As I commented, we cannot do it for unique indices (and I think we did we missed that when we discussed the specification), so it would be a pretty obscure rule.

*not entirely true - indices also have a use independent of the use because a query can read fields from the index even if they are not in the right order, which avoids document lookups. But that's only a relatively small optimization compared to the main use of an index where the order is crucial.

We probably need to look at real-life examples on how we would set up indices with different module combinations.

Reply:
Thanks for the explanation that clears it up a bit 👍